File: DKF-Python/ComputeJacobians.py
import numpy as np

from DeepKalmanFilter.DynJacobian import *
from DeepKalmanFilter.StateJacobian import *

def ComputeJacobians(F, States, Dyn, Inputs, SparseMat, Dynamic, FStateDynInputs, NetParameters):
    """
    Computes the Jacobians of F at the different layers of the net. StateJacobians & DynJacobians are lists of size (1, NetParameters['Layers']) where
    StateJacobians[0] = [] since it is not used during backpropagation.
    """
    # Variables
    Experiment = NetParameters['Experiment']
    Layers = NetParameters['Layers']
    Jacobians = NetParameters['Jacobians']
    N = NetParameters['StateDimension']
    P = NetParameters['HiddenDynamicsDimension']

    # Setup output
    StateJacobians = [None] * Layers
    DynJacobians = [None] * Layers

    if Jacobians == 'Approximated':
        # Approximate Jacobians with finite differences
        if Experiment == '4':
            for Layer in range(1, Layers):
                StateJacobians[Layer] = StateJacobian(F, States[Layer], Dyn, Inputs[:,Layer:Layer+1], SparseMat, FStateDynInputs[Layer], Layer, N, NetParameters)
            
            for Layer in range(Layers):
                DynJacobians[Layer] = DynJacobian(F, States[Layer], Dyn, Inputs[:,Layer:Layer+1], SparseMat, FStateDynInputs[Layer], Layer, N, P[Dynamic-1], NetParameters)
    
    elif Jacobians == 'Algebraic':
        # Set Jacobians to their exact algebraic representation, when possible
        if Experiment == '4':
            for Layer in range(1, Layers):
                # Uncomment and define StateJacobianAlgebraic function when available
                # StateJacobians[Layer] = StateJacobianAlgebraic(F, States[Layer], Dyn, Inputs[Layer], SparseMat, FStateDynInputs[Layer], Layer, N, NetParameters)
                pass
            
            for Layer in range(Layers):
                # Uncomment and define DynJacobianAlgebraic function when available
                # DynJacobians[Layer] = DynJacobianAlgebraic(F, States[Layer], Dyn, Inputs[Layer], SparseMat, FStateDynInputs[Layer], Layer, N, P[Dynamic], NetParameters)
                pass
    
    return StateJacobians, DynJacobians

File: DKF-MATLAB/ComputeJacobians.m
function [StateJacobians,DynJacobians] = ComputeJacobians(F,States,Dyn,Inputs,SparseMat,Dynamic,FStateDynInputs,NetParameters)
%COMPUTEJACOBIANS: Computes the jacobians of F at the different layers of
%the net. StateJacobians & DynJacobians are cells of size (1,NetParameters.Layers) where
%StateJacobians{1} = [] since it is not used during backpropagation.

%Variables
Experiment = NetParameters.Experiment;
Layers = NetParameters.Layers;
Jacobians = NetParameters.Jacobians;
N = NetParameters.StateDimension;
P = NetParameters.HiddenDynamicsDimension;

%Setup output
StateJacobians = cell(1,Layers);
DynJacobians = cell(1,Layers);

if strcmp(Jacobians,'Approximated')
    %Approximate jacobians with finite differences

    if strcmp(Experiment,'1')
        for Layer = 2:Layers
            StateJacobians{Layer} = StateJacobian(F,States{Layer},Dyn,Inputs{Layer},SparseMat,FStateDynInputs{Layer},Layer,N,NetParameters);
        end
        
        for Layer = 1:Layers
            DynJacobians{Layer} = zeros(N,P(Dynamic));
                                 %DynJacobian(F,States{Layer},Dyn,Inputs{Layer},SparseMat,FStateDynInputs{Layer},Layer,N,P(Dynamic),NetParameters);
        end
    end

    if strcmp(Experiment,'2')
        for Layer = 2:Layers
            StateJacobians{Layer} = StateJacobian(F,States{Layer},Dyn,Inputs{Layer},SparseMat,FStateDynInputs{Layer},Layer,N,NetParameters);
        end
        
        for Layer = 1:Layers
            DynJacobians{Layer} = zeros(N,P(Dynamic));
                                 %DynJacobian(F,States{Layer},Dyn,Inputs{Layer},SparseMat,FStateDynInputs{Layer},Layer,N,P(Dynamic),NetParameters);
        end
    end

    if strcmp(Experiment,'3')
        for Layer = 2:Layers
            StateJacobians{Layer} = StateJacobian(F,States{Layer},Dyn,Inputs{Layer},SparseMat,FStateDynInputs{Layer},Layer,N,NetParameters);
        end
        
        for Layer = 1:Layers
            DynJacobians{Layer} = zeros(N,P(Dynamic));
                                 %DynJacobian(F,States{Layer},Dyn,Inputs{Layer},SparseMat,FStateDynInputs{Layer},Layer,N,P(Dynamic),NetParameters);
        end
    end

    if strcmp(Experiment,'4')
        for Layer = 2:Layers
            StateJacobians{Layer} = StateJacobian(F,States{Layer},Dyn,Inputs{Layer},SparseMat,FStateDynInputs{Layer},Layer,N,NetParameters);
        end
        
        for Layer = 1:Layers
            DynJacobians{Layer} = DynJacobian(F,States{Layer},Dyn,Inputs{Layer},SparseMat,FStateDynInputs{Layer},Layer,N,P(Dynamic),NetParameters);
        end
    end

    if strcmp(Experiment,'5')
        for Layer = 2:Layers
            StateJacobians{Layer} = StateJacobian(F,States{Layer},Dyn,Inputs{Layer},SparseMat,FStateDynInputs{Layer},Layer,N,NetParameters);
        end
        
        for Layer = 1:Layers
            DynJacobians{Layer} = DynJacobian(F,States{Layer},Dyn,Inputs{Layer},SparseMat,FStateDynInputs{Layer},Layer,N,P(Dynamic),NetParameters);
        end
    end

    if strcmp(Experiment,'6')
        for Layer = 2:Layers
            StateJacobians{Layer} = StateJacobian(F,States{Layer},Dyn,Inputs{Layer},SparseMat,FStateDynInputs{Layer},Layer,N,NetParameters);
        end
        
        for Layer = 1:Layers
            DynJacobians{Layer} = DynJacobian(F,States{Layer},Dyn,Inputs{Layer},SparseMat,FStateDynInputs{Layer},Layer,N,P(Dynamic),NetParameters);
        end
    end
end

if strcmp(Jacobians,'Algebraic')
    %Set jacobians to their exact algebraic representation, when possible

    if strcmp(Experiment,'1')    
        for Layer = 2:Layers
            %StateJacobians{Layer} = ;
        end

        for Layer = 1:Layers
            %DynJacobians{Layer} = ;
        end
    end

    if strcmp(Experiment,'2')    
        for Layer = 2:Layers
            %StateJacobians{Layer} = ;
        end

        for Layer = 1:Layers
            %DynJacobians{Layer} = ;
        end
    end

    if strcmp(Experiment,'3')    
        for Layer = 2:Layers
            %StateJacobians{Layer} = ;
        end

        for Layer = 1:Layers
            %DynJacobians{Layer} = ;
        end
    end

    if strcmp(Experiment,'4')    
        for Layer = 2:Layers
            %StateJacobians{Layer} = ;
        end

        for Layer = 1:Layers
            %DynJacobians{Layer} = ;
        end
    end

    if strcmp(Experiment,'5')    
        for Layer = 2:Layers
            %StateJacobians{Layer} = ;
        end

        for Layer = 1:Layers
            %DynJacobians{Layer} = ;
        end
    end
end

end

